error/s2n_errno.c

```diff
@@ -205,3 +239,50 @@ int s2n_error_get_type(int error)
 {
     return (error >> S2N_ERR_NUM_VALUE_BITS);
 }
+
+int s2n_error_table_init()
+{
+    /* The hash talbe array uses double size of error array to minimize collisions.
+       It must be initialized with zero before calling hcreate_r.
+    */
+    int table_size = num_of_errors * 2;
+    memset(&error_translation_table, 0, sizeof(struct hsearch_data));
+    int ret = hcreate_r(table_size, &error_translation_table);
+    if (0 == ret) {
+        S2N_ERROR(S2N_ERR_ALLOC);
+    }   
+    
+    for (int i = 0; i < num_of_errors; ++i) {
+        ENTRY e, *ep;
+        char *key = malloc(KEY_LEN);
+        snprintf(key, KEY_LEN, "%x", S2N_ERROR_EN[i].errno_value);
+        e.key = key;
+        e.data = (void *)&S2N_ERROR_EN[i];
+        if (0 == hsearch_r(e, ENTER, &ep, &error_translation_table)) {
+            S2N_ERROR(S2N_ERR_ALLOC);
+        }
+    }
+    
+    error_translation_table_initialized = 0;
+
+    return 0;
+}
+
+void s2n_error_table_cleanup() 
+{
+    if (!error_translation_table_initialized) {
+        return;
+    }
+
+    /* Free key strings */
+    for (int i = 0; i < num_of_errors; ++i) {
+        ENTRY e, *ep;
+        char key[KEY_LEN];
+        snprintf(key, KEY_LEN, "%x", S2N_ERROR_EN[i].errno_value);
+        e.key = key;
+        if (hsearch_r(e, FIND, &ep, &error_translation_table) != 0) {
+            free(ep->key);
+        }
+    }
+    hdestroy_r(&error_translation_table);
+}
```

Comment:
we have a map in utils/s2n_map.c , so if we really need a map/table, there's one we can just use :) But I'm not sure I can see the need. Are the errors already stored at their index offset? Or are they in order? why does performance really matter here? Where does the estimate of a million calls per second come from?

Reply:
> we have a map in utils/s2n_map.c , so if we really need a map/table, there's one we can just use :) 
Yeah. I can switch to s2n_map.

But I'm not sure I can see the need. Are the errors already stored at their index offset? 
I do not thinks so. The error code is catigorized by type (type_start = type << 26).

> Or are they in order? 
I think so. The original code is O(N) like this. Maybe I missed something but open to suggestions. 
for (int i = 0; i < (sizeof(S2N_ERROR_EN) / sizeof(struct s2n_error_translation)); i++) {
        if (S2N_ERROR_EN[i].errno_value == error) {
            return S2N_ERROR_EN[i].str;
        }
}`
  
> Why does performance really matter here? 
That is good question. I do not have measurement right now. In theory, O(1) should be faster than O(N). 

> Where does the estimate of a million calls per second come from?
Our service servels more than million requests per second . In the worst case, we call the 2 APIs for each request. That said, I do not share the exact number here, I can share a metric internally.